### PR TITLE
add connection string parsing

### PR DESCRIPTION
**Pull request by markwolff**
Add connection string parser and parse incoming export config for an instrumentation key and/or ingestion endpoint. If either are specified, existing config is overwritten

### CODE REVIEW

File: src/utils/connectionStringParser.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    if (!connectionString) {
+      return {};
+    }
+
+    const kvPairs = connectionString.split(ConnectionStringParser.FIELDS_SEPARATOR);
+
+    const result: ConnectionString = kvPairs.reduce((fields: ConnectionString, kv: string) => {
+      const kvParts = kv.split(ConnectionStringParser.FIELD_KEY_VALUE_SEPARATOR);
+
+      if (kvParts.length === 2) {
```
Comment (lzchen) [inline src/utils/connectionStringParser.ts:20]: I believe having `Authorization=ikey` present and not present are both valid scenarios.

Comment (lzchen) [inline src/utils/connectionStringParser.ts:20]: I'm not too familiar with js syntax, is this checking that there are only 2 key value pairs, or is it checking that each entry is indeed a pair?

Comment (lzchen) [inline src/utils/connectionStringParser.ts:22]: I believe if split and kvParts is not actually a pair then the connection string is invalid (should fail).

File: src/utils/connectionStringParser.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        const key = kvParts[0].toLowerCase() as ConnectionStringKey;
+        const value = kvParts[1];
+        return { ...fields, [key]: value };
+      }
+      return fields;
+    }, {});
+
+    if (Object.keys(result).length > 0) {
+      // this is a valid connection string, so parse the results
+
```
Comment (lzchen) [inline src/utils/connectionStringParser.ts:41]: Check for `Authorization=ikey`?

File: src/export/exporter.ts [hunk 3]
```diff
@@ -17,13 +18,27 @@ export abstract class AzureMonitorBaseExporter implements BaseExporter {
...
-      this._logger.error('No instrumentation key was provided to the Azure Monitor Exporter');
+    if (!instrumentationKey && !connectionString) {
+      this._logger.error(
+        'No instrumentation key or connection string was provided to the Azure Monitor Exporter',
+      );
+      // @todo: figure out what state the exporter should be left in here
+    }
+
+    if (connectionString) {
+      const parsedConnectionString = ConnectionStringParser.parse(options.connectionString);
```
Comment (xiao-lix) [inline src/export/exporter.ts]: Should the param "connectionString"? Looks like options.connectionString may not exist?

File: src/utils/connectionStringParser.ts (hunk 1 above)
Comment (markwolff) [reply src/utils/connectionStringParser.ts:22]: Just to clarify, do you mean the entire connection string is invalid? Or just the particular kv pair?

File: src/utils/connectionStringParser.ts (hunk 2 above)
Comment (markwolff) [reply src/utils/connectionStringParser.ts:41]: afaik it defaults to `ikey` when not specified. This exporter won't handle any other values of it for now

File: src/export/exporter.ts (hunk 3 above)
Comment (markwolff) [reply src/export/exporter.ts]: fixed

**Review by xiao-lix — APPROVED** [review]

File: src/utils/connectionStringParser.ts (hunk 1 above)
Comment (lzchen) [reply src/utils/connectionStringParser.ts:22]: Both. `foo=bar=test` would split into a kvParts of length 3, so it is not a pair and invalid, and therefore, makes the entire connection string invalid.

File: src/utils/connectionStringParser.ts (hunk 2 above)
Comment (lzchen) [reply src/utils/connectionStringParser.ts:41]: Yes but a user could in theory pass in a value for Authorization that is NOT `ikey` -> `Authorization=invalid` which is technically an invalid connection string.

Comment (markwolff) [reply src/utils/connectionStringParser.ts:41]: makes sense. Added a warning here

File: src/utils/connectionStringParser.ts (hunk 1 above)
Comment (markwolff) [reply src/utils/connectionStringParser.ts:22]: Fixed ptal

**Pull request merged**